### PR TITLE
[Google BigQuery] Added a option to download via Google Cloud Storage

### PR DESCRIPTION
**Pull request by hidekoji**
### Description

Added a integration for Google Cloud Storage

### Checklist

Make sure you have performed following items before submitting this pull request.
If not, please describe the reason.  

- [ ] Add test cases for this fix/enhancement
- [ ] Pass devtools::check()
- [ ] Pass devtools::test() 
- [ ] Tested with Exploratory

### CODE REVIEW

File: R/system.R [hunk 1]
```diff
@@ -508,13 +554,37 @@ saveGoogleBigQueryResultAs <- function(projectId, sourceDatasetId, sourceTableId
 }
 
 #' @export
-executeGoogleBigQuery <- function(project, sqlquery, destination_table, page_size = 10000, max_page = 10, write_disposition = "WRITE_TRUNCATE", tokenFileId){
+executeGoogleBigQuery <- function(project, sqlquery, destination_table, page_size = 10000, max_page = 10, write_disposition = "WRITE_TRUNCATE", tokenFileId, bucketProjectId, bucket=NULL, folder){
```
Comment (hideaki) [inline R/system.R]: Comments on parameters would help.

File: R/system.R [hunk 2]
```diff
@@ -508,13 +554,37 @@ saveGoogleBigQueryResultAs <- function(projectId, sourceDatasetId, sourceTableId
...
   if(!requireNamespace("bigrquery")){stop("package bigrquery must be installed.")}
   if(!requireNamespace("GetoptLong")){stop("package GetoptLong must be installed.")}
+  if(!requireNamespace("stringr")){stop("package stringr must be installed.")}
 
   token <- getGoogleTokenForBigQuery(tokenFileId)
-  bigrquery::set_access_cred(token)
-  bigrquery::query_exec(GetoptLong::qq(sqlquery), project = project, destination_table = destination_table, page_size = page_size, max_page = max_page, write_disposition = write_disposition)
+  df <- NULL
+  if(!is.na(bucket) && bucket != ""){
```
Comment (hideaki) [inline R/system.R]: that when this if statement is true, export is done through cloud storage should be worth commenting.

File: R/system.R [hunk 3]
```diff
@@ -495,6 +498,49 @@ getDataFromGoogleBigQueryTable <- function(project, dataset, table, page_size =
                  table_info = NULL, max_pages = max_page)
 }
 
+#' API to extract datafrom google BigQuery table to Google Storage
+#' @export
+extractDataFromGoogleBigQueryTableToStorage <- function(project, dataset, table, destinationUri, tokenFileId){
```
Comment (hideaki) [inline R/system.R]: About this function name, in other part of the code, google cloud storage is referred to as "CloudStorage", but here only as "Storage", which looks a little inconsistent.
Also, thought that "Table" of "BigQueryTable" is kind of redundant and can be omitted.

File: R/system.R [hunk 4]
```diff
@@ -495,6 +498,49 @@ getDataFromGoogleBigQueryTable <- function(project, dataset, table, page_size =
                  table_info = NULL, max_pages = max_page)
 }
 
+#' API to extract datafrom google BigQuery table to Google Storage
```
Comment (hidekoji) [inline R/system.R]: ah, it's `data from`

File: R/system.R (hunk 3 above)
Comment (hidekoji) [inline R/system.R]: Changed as suggested

File: R/system.R [hunk 5]
```diff
@@ -495,6 +498,49 @@ getDataFromGoogleBigQueryTable <- function(project, dataset, table, page_size =
...
+#' @export
+extractDataFromGoogleBigQueryTableToStorage <- function(project, dataset, table, destinationUri, tokenFileId){
+  if(!requireNamespace("bigrquery")){stop("package bigrquery must be installed.")}
+  token <- getGoogleTokenForBigQuery(tokenFileId)
+  bigrquery::set_access_cred(token)
+
+  job <- bigrquery::insert_extract_job(project, dataset, table, destinationUri,
```
Comment (hidekoji) [inline R/system.R]: Added comment

File: R/system.R [hunk 6]
```diff
@@ -495,6 +498,49 @@ getDataFromGoogleBigQueryTable <- function(project, dataset, table, page_size =
...
+                                print_header=TRUE, field_delimiter=",", destination_format="CSV", compression="GZIP")
+  job <- bigrquery::wait_for(job)
+  job
+}
+
+#' API to download datafrom Google Storage to client and create a df from it
```
Comment (hidekoji) [inline R/system.R]: It's `data from`

File: R/system.R [hunk 7]
```diff
@@ -495,6 +498,49 @@ getDataFromGoogleBigQueryTable <- function(project, dataset, table, page_size =
...
+#' @export
+downloadDataFromGoogleCloudStorage <- function(bucket, folder, download_dir, tokenFileId){
+  if(!requireNamespace("googleCloudStorageR")){stop("package googleCloudStorageR must be installed.")}
+  if(!requireNamespace("googleAuthR")){stop("package googleAuthR must be installed.")}
+  token <- getGoogleTokenForBigQuery(tokenFileId)
+  googleAuthR::gar_auth(token = token)
+  googleCloudStorageR::gcs_global_bucket(bucket)
+  objects <- googleCloudStorageR::gcs_list_objects()
+  # set buckt
```
Comment (hidekoji) [inline R/system.R]: fixed typo

File: R/system.R [hunk 8]
```diff
@@ -495,6 +498,49 @@ getDataFromGoogleBigQueryTable <- function(project, dataset, table, page_size =
...
+  googleCloudStorageR::gcs_global_bucket(bucket)
+  objects <- googleCloudStorageR::gcs_list_objects()
+  lapply(objects$name, function(name){
+    if(stringr::str_detect(name,stringr::str_c(folder, "/"))){
+      googleCloudStorageR::gcs_get_object(name, saveToDisk = str_c(download_dir, "/", stringr::str_replace(name, stringr::str_c(folder, "/"),"")))
+      googleCloudStorageR::gcs_delete_object(name, bucket = bucket)
+    }
+  });
+  files <- list.files(path=download_dir, pattern = ".gz");
+  df <- lapply(files, function(file){readr::read_csv(stringr::str_c(download_dir, "/", file))}) %>% dplyr::bind_rows()
```
Comment (hidekoji) [inline R/system.R]: added comment

File: R/system.R (hunk 1 above)
Comment (hidekoji) [inline R/system.R]: Added comments around parameter

File: R/system.R (hunk 2 above)
Comment (hidekoji) [inline R/system.R]: added comment

**Pull request merged**